### PR TITLE
Make scripted search templates work with new mediaType from XContentType.JSON

### LINKED ISSUE #66986: Test failing due to wrong content_type - 10_basic/Verify that we can still find things with the template
introduced by https://github.com/elastic/elasticsearch/pull/65500

**Build scan**:
https://gradle-enterprise.elastic.co/s/cypeuyt4db6ow/tests/:qa:rolling-upgrade:v7.12.0%23oneThirdUpgradedTest/org.elasticsearch.upgrades.UpgradeClusterClientYamlTestSuiteIT/test%20%7Bp0=mixed_cluster%2F10_basic%2FVerify%20that%20we%20can%20still%20find%20things%20with%20the%20template%7D#1

**Repro line**:
```
./gradlew ':qa:rolling-upgrade:v7.12.0#oneThirdUpgradedTest' -Dtests.class="org.elasticsearch.upgrades.UpgradeClusterClientYamlTestSuiteIT" -Dtests.method="test {p0=mixed_cluster/10_basic/Verify that we can still find things with the template}" -Dtests.seed=CCF20BC12765F837 -Dtests.security.manager=true -Dtests.bwc=true -Dtests.locale=he-IL -Dtests.timezone=PRC -Druntime.java=11 -Dtests.rest.suite=mixed_cluster	

```
**Reproduces locally?**:
yes

**Applicable branches**:

**Failure history**:

**Failure excerpt**:
```
"type" : "general_script_exception", |  
-- | --
  | "reason" : "Failed to compile stored script [test_search_template] using lang [mustache]", |  
  | "stack_trace" : "org.elasticsearch.script.GeneralScriptException: Failed to compile stored script [test_search_template] using lang [mustache] |  
  | at org.elasticsearch.script.ScriptCache.compile(ScriptCache.java:121) |  
  | at org.elasticsearch.script.ScriptService.compile(ScriptService.java:339) |  
  | at org.elasticsearch.script.mustache.TransportSearchTemplateAction.convert(TransportSearchTemplateAction.java:100) |  
  | at org.elasticsearch.script.mustache.TransportSearchTemplateAction.doExecute(TransportSearchTemplateAction.java:69) |  
  | at org.elasticsearch.script.mustache.TransportSearchTemplateAction.doExecute(TransportSearchTemplateAction.java:48) |  
  | at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:88) |  
  | at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:64) |  
  | at org.elasticsearch.tasks.TaskManager.registerAndExecute(TaskManager.java:175) |  
  | at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:108) |  
  | at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:88) |  
  | at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:386) |  
  | at org.elasticsearch.script.mustache.RestSearchTemplateAction.lambda$prepareRequest$1(RestSearchTemplateAction.java:77) |  
  | at org.elasticsearch.rest.BaseRestHandler.handleRequest(BaseRestHandler.java:101) |  
  | at org.elasticsearch.rest.RestController.dispatchRequest(RestController.java:268) |  
  | at org.elasticsearch.rest.RestController.tryAllHandlers(RestController.java:353) |  
  | at org.elasticsearch.rest.RestController.dispatchRequest(RestController.java:198) |  
  | at org.elasticsearch.http.AbstractHttpServerTransport.dispatchRequest(AbstractHttpServerTransport.java:336) |  
  | at org.elasticsearch.http.AbstractHttpServerTransport.handleIncomingRequest(AbstractHttpServerTransport.java:401) |  
  | at org.elasticsearch.http.AbstractHttpServerTransport.incomingRequest(AbstractHttpServerTransport.java:318) |  
  | at org.elasticsearch.http.netty4.Netty4HttpRequestHandler.channelRead0(Netty4HttpRequestHandler.java:42) |  
  | at org.elasticsearch.http.netty4.Netty4HttpRequestHandler.channelRead0(Netty4HttpRequestHandler.java:28) |  
  | at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) |  
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) |  
  | at org.elasticsearch.http.netty4.Netty4HttpPipeliningHandler.channelRead(Netty4HttpPipeliningHandler.java:58) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) |  
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) |  
  | at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) |  
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) |  
  | at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) |  
  | at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) |  
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) |  
  | at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) |  
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) |  
  | at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) |  
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) |  
  | at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324) |  
  | at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) |  
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) |  
  | at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) |  
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) |  
  | at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) |  
  | at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) |  
  | at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) |  
  | at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) |  
  | at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) |  
  | at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) |  
  | at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714) |  
  | at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:615) |  
  | at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:578) |  
  | at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) |  
  | at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) |  
  | at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) |  
  | at java.base/java.lang.Thread.run(Thread.java:834) |  
  | Caused by: java.lang.IllegalArgumentException: No encoder found for MIME type [application/json; charset=UTF-8] |  
  | at org.elasticsearch.script.mustache.CustomMustacheFactory.createEncoder(CustomMustacheFactory.java:92) |  
  | at org.elasticsearch.script.mustache.CustomMustacheFactory.<init>(CustomMustacheFactory.java:73) |  
  | at org.elasticsearch.script.mustache.MustacheScriptEngine.createMustacheFactory(MustacheScriptEngine.java:96) |  
  | at org.elasticsearch.script.mustache.MustacheScriptEngine.compile(MustacheScriptEngine.java:75) |  
  | at org.elasticsearch.script.ScriptCache.lambda$compile$0(ScriptCache.java:110) |  
  | at org.elasticsearch.common.cache.Cache.computeIfAbsent(Cache.java:435) |  
  | at org.elasticsearch.script.ScriptCache.compile(ScriptCache.java:99) |  
  | ... 70 more
```

### PR DESCRIPTION
**Pull request by pgomulka**
Stored scripts can have `content_type` option set, however when empty they default to `XContentType.JSON#mediaType()`. Commit  5e74f79 has changed this in master (ES8) method to return `application/json;charset=utf-8` (previously `application/json; charset=UTF-8`)
This means that when upgrading ES from version 7 to 8 stored script will fail when being used as the encoder is being matched with string equality (map key) https://github.com/elastic/elasticsearch/blob/5e74f79e22c4df1edaa224219453c01215cbabee/modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java#L62-L66

This commit address this by adding back (in addition) the old `application/json; charset=UTF-8` into the encoders map.

closes #66986

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/util/Maps.java [hunk 1]
```diff
@@ -71,6 +72,26 @@
         return Stream.concat(map.entrySet().stream().filter(k -> key.equals(k.getKey()) == false), Stream.of(entry(key, value)))
                 .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/common/util/Maps.java:73]: possibly we could optimise and return the original map if the key&value are the same as already existing in the original map.

File: server/src/main/java/org/elasticsearch/script/Script.java [hunk 2]
```diff
@@ -531,7 +532,10 @@ public Script(ScriptType type, String lang, String idOrCode, Map<String, String>
 
         if (type == ScriptType.INLINE) {
             this.lang = Objects.requireNonNull(lang);
-            this.options = Collections.unmodifiableMap(Objects.requireNonNull(options));
+            // making sure if content_type was application/json, application/json; charset=UTF-8 (note space and casing)
+            // it will be treated as the XContentType.JSON.mediaType() -> application/json;charset=utf-8
+            this.options = Maps.copyMapWithModifiedEntryWhenPresent(Collections.unmodifiableMap(options), CONTENT_TYPE_OPTION,
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/script/Script.java:537]: we don't document much about this `content_type` field in scripts. I suspect users might still send the `application/json; charset=UTF-8`. This line is meant to make it backwards compatible with that value and always set `application/json;charset=utf-8` as an option when `application/json*` was used.

File: server/src/main/java/org/elasticsearch/script/Script.java [hunk 3]
```diff
@@ -633,7 +637,8 @@ public XContentBuilder toXContent(XContentBuilder builder, Params builderParams)
         String contentType = options == null ? null : options.get(CONTENT_TYPE_OPTION);
 
         if (type == ScriptType.INLINE) {
-            if (contentType != null && builder.contentType().mediaType().equals(contentType)) {
+            // treating application/json;charset=utf-8 the same as application/json
+            if (contentType != null && (contentType.startsWith(builder.contentType().mediaTypeWithoutParameters()))) {
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/script/Script.java:641]: related to https://github.com/elastic/elasticsearch/pull/67677/files#r560424309 - although possibly not needed as we will always save `XContentType.JSON.mediaType()`
but can `options.content_type` have a different value than JSON? like YAML?
also - how do I trigger this code? when `Script#toXContent` is used?

@martijnvg I was advised that you might know about this?

**Comment by pgomulka** [comment]
I removed the Script.java related changes to a separate PR

**Review by jdconrad — APPROVED** [review]
LGTM. Thanks for working on this!

**Review by martijnvg — APPROVED** [review]
LGTM

**Pull request merged**